### ISSUE TITLE
Add a new standard protocol for disabling the behavior of accepting all memory at ExitBootServices (Bugzilla Bug 3987)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9158]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 3987
Date: 2022-07-12T19:14:56+00:00
From: Dionna Glaze &lt;<dionnaglaze>&gt;
To:   Dionna Glaze &lt;<dionnaglaze>&gt;
CC:   @ardbiesheuvel, @ruleof2, @lgao4, @jyao1, @mdroth, @mxu9

Last updated: 2023-01-31T19:50:11+00:00

**Comment by tianocore-issues** [comment]
## Comment 19233
Date: 2022-07-12 19:14:56 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

- Industry Specification: UEFI Specification
- Target OS: ---
- Bugzilla Assignee(s): Dionna Glaze &lt;<dionnaglaze>&gt;

# Title: GetMemoryMapEx

# Status: Draft

# Document: UEFI Specification Version 2.10

# License

SPDX-License-Identifier: CC-BY-4.0

# Submitter: [TianoCore Community](https://www.tianocore.org)

# Summary of the change

Adds a new memory allocation boot service GetMemoryMapEx that carries a memory feature support bitmap.

# Benefits of the change

The addition of unaccepted memory in v2.9 makes the assumption that support of a technology (e.g., Intel TDX, AMD SEV-SNP) is paired with support of unaccepted memory in the guest OS. This is not the case. Linux has support for SEV-SNP without support for accepting unaccepted memory. By adding this boot service, the guest OS can inform the UEFI whether it has support for unaccepted memory or not so that in the case that unaccepted memory is not supported, the UEFI can accept all memory and classify unaccepted memory as system memory before returning a memory map.

# Impact of the change

Virtual machines that run SEV-SNP will enjoy the benefits of a faster boot, since accepting all memory before booting the guest OS can take a significant amount of time for VMs with a large memory footprint.

# Detailed description of the change [normative updates]

Addition to end of Section 4.4 EFI Boot services table

 EFI_GET_MEMORY_MAP_EX GetMemoryMapEx; // UEFI 2.10+

Addition to Section 7.2

EFI_BOOT_SERVICES.GetMemoryMapEx()

Summary
Returns the current memory map subject to feature support.

Prototype

typedef EFI_STATUS (EFIAPI *EFI_GET_MEMORY_MAP_EX) (
  IN OUT UINTN *MemoryMapSize,
  IN         EFI_MEMORY_MAP_FEATURES_TYPE *MemoryMapFeatures,
  OUT EFI_MEMORY_DESCRIPTOR *MemoryMap,
  OUT UINTN *MapKey,
  OUT UINTN *DescriptorSize,
  OUT UINT32 *DescriptorVersion
  );

Parameters

MemoryMapSize

A pointer to the size, in bytes, of the MemoryMap buffer. On input, this is the size of the buffer allocated by the caller. On output, it is the size of the buffer returned by the firmware if the buffer was large enough, or the size of the buffer needed to contain the map if the buffer was too small.

MemoryMapFeatures

A pointer to the description of the caller&apos;s support for particular memory features. The description is a sized struct to permit future extension. If NULL, then all features are considered supported. See “Related Definitions.”

MemoryMap

A pointer to the buffer in which firmware places the current memory map. The map is an array of EFI_MEMORY_DESCRIPTORs, which is defined in the GetMemoryMap() function description.

MapKey

A pointer to the location in which firmware returns the key for the current memory map. DescriptorSize A pointer to the location in which firmware returns the size, in bytes, of an individual EFI_MEMORY_DESCRIPTOR.

DescriptorVersion

A pointer to the location in which firmware returns the version number associated with the EFI_MEMORY_DESCRIPTOR. See “Related Definitions.”

EFI_SUCCESS The requested memory pages were freed.
EFI_NOT_FOUND The requested memory pages were not allocated with AllocatePages(). EFI_INVALID_PARAMETER Memory is not a page-aligned address or Pages is invalid.

Related Definitions

//*******************************************************
//EFI_MEMORY_MAP_FEATURES_TYPE
//*******************************************************
typedef struct {
  UINT32 Size;
  UINT64 FeatureBitmap0;
} EFI_MEMORY_MAP_FEATURES_TYPE;

//*******************************************************
// Acceptable values for FeatureBitmap0 in EFI_MEMORY_MAP_FEATURES_TYPE
//*******************************************************
#define EFI_MEMORY_MAP_FEATURE0_UNACCEPTED_MEMORY 1

Size

The size of this struct. If less than the size of the struct, then the remaining data are defined as 0. If larger than the size of the struct, then all data beyond is considered unsupported optional feature data and is defined as 0.

FeatureBitmap0

A bitmap of features that the caller supports. Must be a bitwise OR of EFI_MEMORY_MAP_FEATURE0_* values.

**Comment by tianocore-issues** [comment]
## Comment 19254
Date: 2022-07-18 12:27:18 +0000
From: @jyao1

To help me understand the proposal, would you please clarify couple of things?

1) When you say &quot;guest OS&quot;, are you referring to &quot;OS loader&quot;, or &quot;OS kernel&quot;, or the combination?

2) If it means combination, do you only want to support all below cases, or a subset?

+------+-----------+-----------+
| case | OS loader | OS kernel |
+------+-----------+-----------+
| A    | NO        | NO        |
+------+-----------+-----------+
| B    | NO        | YES       |
+------+-----------+-----------+
| C    | YES       | NO        |
+------+-----------+-----------+
| D    | YES       | YES       |
+------+-----------+-----------+
where YES == knows unaccepted-memory, NO == does not know unaccepted-memory.

3) Would you please give a flow on how the GetMemoryMap() or GetMemoryMapEx() is called in those 4 cases?

NOTE: Both OS loader and OS kernel may call GetMemoryMap() or GetMemoryMapEx().
Some BIOS code may also call GetMemoryMap() internally.

4) When does ACCEPT-ALL happen in above 4 cases?

**Comment by tianocore-issues** [comment]
## Comment 19255
Date: 2022-07-18 12:50:49 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; 1) When you say &quot;guest OS&quot;, are you referring to &quot;OS loader&quot;, or &quot;OS kernel&quot;, or the combination?

I mean the combination.

&gt; 2) If it means combination, do you only want to support all below cases, or a subset? [2-bit table naming 0-3 as A,B,C,D and truth value := GetMemoryMapEx enlightened]

I think A, C, and D are what makes sense to support, although there&apos;s no reason to forbid all combinations. If the OS loader calls GetMemoryMap before (say) Linux calls GetMemoryMapEx, the all memory would be accepted at that point and there wouldn&apos;t be any unaccepted memory to utilize.

&gt; 3) Would you please give a flow on how the GetMemoryMap() or GetMemoryMapEx() is called in those 4 cases?
&gt; NOTE: Both OS loader and OS kernel may call GetMemoryMap() or GetMemoryMapEx().
Some BIOS code may also call GetMemoryMap() internally.

The proposed change I have to EDK2 changes all calls to GetMemoryMap with GetMemoryMapEx with a NULL features pointer to not resolve unsupported features&quot;.

GetMemoryMap becomes a call to GetMemoryMapEx with a zeroed feature struct to state all memory features are unsupported. Every call to GetMemoryMap thus tries to change unaccepted memory to accepted memory, but only the first time is it not a no-op.

&gt; 4) When does ACCEPT-ALL happen in above 4 cases?

Any call to GetMemoryMapEx with a non-null features struct that doesn&apos;t set the unaccepted memory feature bit will accept all unaccepted memory and convert those memory ranges to conventional system memory.

A. Thus, an unenlightened OS loader that builds an E820 map for the OS that calls GetMemoryMap instead of GetMemoryMapEx will accept all memory. The guest OS doesn&apos;t see the unaccepted memory ranges.

B. Same as A.

C. The OS loader doesn&apos;t accept all memory. The guest OS doesn&apos;t know about GetMemoryMapEx, so it does accept all memory at the time it calls GetMemoryMap.

D. Both OS loader and OS use GetMemoryMapEx with the unaccepted memory feature bit so there is no accept-all event.

**Comment by tianocore-issues** [comment]
## Comment 19260
Date: 2022-07-18 21:00:16 +0000
From: @jyao1

It seems the pattern you want to build is: The first GetMemoryMap() from either OS loader or OS kernel will cause ACCEPT-ALL.

The assumption is that: The module that does not know Unaccepted-Memory must be in UEFI boot environment and must call GetMemoryMap() to trigger the ACCEPT-ALL.

I am not sure if that is right assumption, because that will force the kernel to be UEFI-aware. What if the kernel is not?

Case A, B, and D are still good.
Case C is problematic then.

Apparently, I forget to add case E and F, which are kernel direct boot.
+------+-----------+-----------+
| case | OS loader | OS kernel |
+------+-----------+-----------+
| E    | --        | NO        |
+------+-----------+-----------+
| F    | --        | YES       |
+------+-----------+-----------+
where &quot;--&quot; means the OS loader is absent, the VBIOS boot kernel directly.


I am thinking other simplification models.

1) We restrict the decision from *all boot image* to *first boot image*.

The first booted image decides the ACCEPT-ALL or not. (Not all booted images).

That means in case A,B,D, we only rely on boot loader to make right decision. The Boot loader may need figure out if the kernel supports Unaccepted memory - this can resolve the non-UEFI kernel issue.

Case C is unsupported.

In case E,F, we can rely on kernel. What if the kernel is non-UEFI?

2) We change the decision from *function call* to *capability report*.

Instead of calling a function from guest OS, we may let guest OS self-report the capability. (i.e. from a bit in kernel header).

This model is simple and more natural from my perspective.

2.1) The TDVF can detect it, if the cap bit is detected from boot image directly. It does not care the guest is UEFI or not.

Case C is still a problem, because the TDVF may not have chance to touch the kernel directly. All TDVF can see is the first boot image.

2.2) The VMM can detect it from cap bit or from a manifest, which is more flexible.

The VMM can detect it and pass the info to TDVF. Because the VMM is untrusted, it will be measured to RTMR for attestation later. This is a potential attack point, because VMM may lie - when the guest does not support, VMM says it supports. The consequence is: the TDVF wrongly configures the unaccepted memory and cause problem in the kernel (hang or low memory). -- I treat it as deny-of-server (DOS) attack, but DOS is out of scope of confidential computing, because the VMM can DOS anyway - jump to wrong kernel or configure lower memory.

**Comment by tianocore-issues** [comment]
## Comment 19261
Date: 2022-07-18 21:20:48 +0000
From: @jyao1

I think the third way:

3) Use UEFI &quot;OsIndications&quot; variable. Ref: UEFI specification 2.9, section 8.5.4.

The pattern is:
* BIOS reports the capability via - &quot;OsIndicationsSupported&quot; variable.
* OS reports the enable/disable via - &quot;OsIndications&quot; non-volatile variable.

However, it has two limitations / assumptions:
* The guest OS must be UEFI aware.
* The guest BIOS must support non-volatile storage.

**Comment by tianocore-issues** [comment]
## Comment 19266
Date: 2022-07-19 13:50:51 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Case C is where we&apos;re at with SEV-SNP right now. We&apos;d have the enlightened UEFI but a non-enlightened guest OS.

It sounds like the solution you&apos;re suggesting is a combination for UEFI and non-UEFI kernels, but a solution for a non-UEFI kernel could still work for a UEFI kernel it seems.

I think requiring a capability bit that is in the kernel binary for the bootloader or direct boot path to utilize would be best.

For linux, we can use the `setup_data` linked list by the boot header at offset 0x01f1 of the kernel binary: https://www.kernel.org/doc/html/latest/x86/boot.html to have a memory feature bitmap like this bug suggests for GetMemoryMapEx.

**Comment by tianocore-issues** [comment]
## Comment 19267
Date: 2022-07-19 13:57:08 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Unfinished comment accidentally sent. I meant to follow up with a question: given a communication channel for the OS and bootloader, we still don&apos;t solve the &quot;C&quot; case if we change the boot service to be an &quot;explicitly ask for accept all&quot; (supports B, not C) instead of &quot;explicitly ask to not accept all&quot; model (supports C, not B).

To solve &quot;B&quot;, I think we need to incorporate your idea of the OsIndications variable as well, and have it override the feature support passed in to GetMemoryMapEx.

**Comment by tianocore-issues** [comment]
## Comment 19312
Date: 2022-07-20 05:24:51 +0000
From: @jyao1

With this bit definition, we can resolve case A,B,D,E,F.

Then we don&apos;t need this GetMemoryMapEx API, right?
We just need a &quot;new mechanism&quot; to let the boot loader explicitly request to accept all memory, if it detects the kernel has no capability. That resolves the case C.

The UEFI variable is not required, because the accept should happen in current boot, immediately.

The new mechanism could be a UEFI protocol to provide AcceptMemory() API. (optional, and only for confidential computing). It seems more natural to me.

**Comment by tianocore-issues** [comment]
## Comment 19314
Date: 2022-07-20 15:21:47 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

I don&apos;t see how we resolve (A) with just this bit if there&apos;s a bootloader that&apos;s required to accept all memory if the OS doesn&apos;t support unaccepted memory. We&apos;d still end up with the UEFI creating unaccepted memory that nothing then accepts.

If we go through a typical path of UEFI -&gt; shim -&gt; GRUB2 -&gt; Linux, then we simply can&apos;t get the information to the UEFI that all memory should be accepted because neither GRUB2 nor Linux support unaccepted memory.

**Comment by tianocore-issues** [comment]
## Comment 19315
Date: 2022-07-20 15:41:20 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

I think the combination of an OsIndications bit and the self-report kernel header bit could be a full solution, but a new OsIndications bit requires a spec change, correct?

A. No OsIndications bit, accept all.
B. OsIndications bit set on second boot, so don&apos;t accept all.
C. No OsIndications bit, accept all.
D. OsIndications bit set on second boot, so don&apos;t accept all.
E. VMM inspects kernel binary for bit, doesn&apos;t create unaccepted memory regions for e820.
F. VMM inspects kernel binary for bit, creates unaccepted memory regions for e820.

**Comment by tianocore-issues** [comment]
## Comment 19325
Date: 2022-07-22 15:15:32 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

For the OsIndication bit idea, I have a PoC that I&apos;ve tested internally.

EDK2 changes

https://github.com/deeglaze/edk2/tree/osindicationaccept

Linux changes (with Kirill&apos;s patch set and Brijesh&apos;s SEV-SNP extension):

https://github.com/deeglaze/amdese-linux/commit/f1c3db19d97d1851fec39d0d1de97fb11143f214

The single bit here works alright and is pretty simple to use.

**Comment by tianocore-issues** [comment]
## Comment 19390
Date: 2022-08-11 12:12:33 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Any other comments on how we can solve this upgrade path without external information about OS images?

**Comment by tianocore-issues** [comment]
## Comment 19480
Date: 2022-09-06 21:37:04 +0000
From: @lgao4

Jiewen: have you any other comments?

Dionna Glaze: can you help title to match new change?

**Comment by tianocore-issues** [comment]
## Comment 19512
Date: 2022-09-07 04:22:53 +0000
From: @jyao1

&gt; a new OsIndications bit requires a spec change, correct?

That is correct. Any contract between BIOS and OS need UEFI spec change.
But I think this is a minor update.

**Comment by tianocore-issues** [comment]
## Comment 19513
Date: 2022-09-07 05:00:22 +0000
From: @ardbiesheuvel

(In reply to Dionna Glaze from comment #9)
&gt; I think the combination of an OsIndications bit and the self-report kernel
&gt; header bit could be a full solution, but a new OsIndications bit requires a
&gt; spec change, correct?
&gt; 
&gt; A. No OsIndications bit, accept all.
&gt; B. OsIndications bit set on second boot, so don&apos;t accept all.
&gt; C. No OsIndications bit, accept all.
&gt; D. OsIndications bit set on second boot, so don&apos;t accept all.
&gt; E. VMM inspects kernel binary for bit, doesn&apos;t create unaccepted memory
&gt; regions for e820.
&gt; F. VMM inspects kernel binary for bit, creates unaccepted memory regions for
&gt; e820.

What I am missing here is the justification for supporting all these combinations in what amounts to a highly bespoke and restricted confidential VM scenario.

The attestation and memory encryption pieces of confidential computing rely heavily on firmware abstractions. Is it really a priority, and worth obfuscating the picture, to support a boot mode where a legacy BIOS boots directly into a Linux kernel without awareness of the EFI firmware at all? If so, could you elaborate on how such a deployment still makes meaningful use of the features that confidential VMs provide?

In the light of this, I think it may have been a mistake to use a completely new memory type for unaccepted memory as opposed to a memory attribute.

As for the use of OsIndications: I think this is a reasonable and idiomatic way of using this feature. I do wonder if we could &apos;abuse&apos; the fact that GetMemoryMap()+ExitBootServices() are documented as potentially needing to be invoked twice, and accept all memory at ExitBootServices() time if the OsIndication has not been set by the EFI stub (as opposed to the OS proper). ExitBootServices() would accept all memory and return an error, forcing the caller to call GetMemoryMap() again to obtain the updated version before calling EBS() again. If the OsIndication has been set, things proceed as usual.

**Comment by tianocore-issues** [comment]
## Comment 19524
Date: 2022-09-07 19:37:20 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; What I am missing here is the justification for supporting all these combinations in what amounts to a highly bespoke and restricted confidential VM scenario.

We have customers in each category is the short of it. There&apos;s demand for supporting all these cases.

&gt; The attestation and memory encryption pieces of confidential computing rely heavily on firmware abstractions. Is it really a priority, and worth obfuscating the picture, to support a boot mode where a legacy BIOS boots directly into a Linux kernel without awareness of the EFI firmware at all? If so, could you elaborate on how such a deployment still makes meaningful use of the features that confidential VMs provide?

I&apos;m not sure I understand your question. Canonical for example ships their Ubuntu virtual machine images with GRUB2 installed, so it may call ExitBootServices before Linux can. I&apos;m not talking about Linux having no understanding of the EFI. I&apos;m talking about supporting a clean path where Linux might not be compiled with CONFIG_UNSUPPORTED_MEMORY=y, which is an allowed configuration for both SEV-SNP and TDX.

&gt; As for the use of OsIndications: I think this is a reasonable and idiomatic way of using this feature. I do wonder if we could &apos;abuse&apos; the fact that GetMemoryMap()+ExitBootServices() are documented as potentially needing to be invoked twice, and accept all memory at ExitBootServices() time if the OsIndication has not been set by the EFI stub (as opposed to the OS proper). 
ExitBootServices() would accept all memory and return an error, forcing the caller to call GetMemoryMap() again to obtain the updated version before calling EBS() again. If the OsIndication has been set, things proceed as usual.

I&apos;ve implemented this and it doesn&apos;t work without a reboot when using GRUB2 to boot Linux. It calls ExitBootServices before launching Linux. We&apos;d have to do some kind of inspection in GRUB2 to check if the kernel supports unaccepted memory, or add a grub configuration thingy and have it perform the EFI variable dance. That kernel inspection is something I know you&apos;ve NACK&apos;d explicitly. GRUB2 configuration is kind of this problem all over again, so I feel a bit stuck.

**Comment by tianocore-issues** [comment]
## Comment 19525
Date: 2022-09-07 19:42:08 +0000
From: @mxu9

(In reply to Dionna Glaze from comment #9)
&gt; I think the combination of an OsIndications bit and the self-report kernel
&gt; header bit could be a full solution, but a new OsIndications bit requires a
&gt; spec change, correct?
&gt; 
&gt; A. No OsIndications bit, accept all.
&gt; B. OsIndications bit set on second boot, so don&apos;t accept all.
&gt; C. No OsIndications bit, accept all.
&gt; D. OsIndications bit set on second boot, so don&apos;t accept all.
&gt; E. VMM inspects kernel binary for bit, doesn&apos;t create unaccepted memory
&gt; regions for e820.
&gt; F. VMM inspects kernel binary for bit, creates unaccepted memory regions for
&gt; e820.

It seems OsIndications is ignored in E/F, right? VMM inspects kernel binary for the bit and then what VMM does? Does VMM pass the capability of unaccept memmory of the kernel to FW in some way, for example, fw_cfg (QEMU talks with FW) or TD-Hob (VMM pass information in TD-Hob to TDVF)? If this is the case, I am thinking E/F can handle A/B/C/D as well.

Another question is that it seems the discussion is focused on whether the kernel support unaccept-mem or not. Shall we also consider the minimal amount of the memory that kernel needs. So that FW can accept that amount of memory then let OS handle the left unaccept-mem.

**Comment by tianocore-issues** [comment]
## Comment 19526
Date: 2022-09-07 19:51:44 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

The E/F options were speculative as to whether the OS community (Linux in my case) would accept baking in this kind of configuration information directly into the binary. Ard was firmly against this https://lore.kernel.org/linux-mm/CAMj1kXHEc=vEt=CtfdiPEsUe2i8QogAi+jvtY6h1awo7GZ-nRg@mail.gmail.com/

My original idea about E/F would be to inspect the binary before running anything, then set up a QemuFwCfg file opt/ovmf/AcceptAllMemory if the binary was missing the appropriate information. We&apos;d use that file to determine in PEI whether to accept all memory or allow creation of unaccepted memory.

With inspection, we couldn&apos;t handle something as complicated (impossible?) as finding and analyzing a grub.cfg file to determine in the kernel that will be booted will have the appropriate support information included. Boot loaders make this hard, since they can call ExitBootServices before the guest kernel has a chance to start and indicate the right support bit. That is, if we use the OsIndications solution over the GetMemoryMapEx solution.

**Comment by tianocore-issues** [comment]
## Comment 19527
Date: 2022-09-07 21:33:21 +0000
From: @mxu9

(In reply to Dionna Glaze from comment #17)
&gt; The E/F options were speculative as to whether the OS community (Linux in my
&gt; case) would accept baking in this kind of configuration information directly
&gt; into the binary. Ard was firmly against this
&gt; https://lore.kernel.org/linux-mm/
&gt; CAMj1kXHEc=vEt=CtfdiPEsUe2i8QogAi+jvtY6h1awo7GZ-nRg@mail.gmail.com/

So is self-report kernel binary bit still an option?

**Comment by tianocore-issues** [comment]
## Comment 19529
Date: 2022-09-07 22:17:08 +0000
From: @jyao1

The only downside of OsIndicator is that: This solution assumes that the firmware has OS-writable non-volatile storage.

Maybe not a problem for UEFI, since UEFI supports SetVariable().

But it might be a problem for non-UEFI system.

**Comment by tianocore-issues** [comment]
## Comment 19534
Date: 2022-09-08 11:52:51 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; So is self-report kernel binary bit still an option?

It&apos;s unlikely, but maybe we can convince Linux upstream that it&apos;s really necessary. There&apos;s a mechanism for it in the boot protocol called kernel_info.

&gt; Shall we also consider the minimal amount of the memory that kernel needs. So that FW can accept that amount of memory then let OS handle the left unaccept-mem.

We could add a 64-bit entry in kernel_info called required_accepted_memory and only have it be non-zero if the kernel supports unaccepted memory.

Ard, what do you think about this? I think since it involves necessary information for booting (accepting the right amount of memory), this should fit in the boot protocol.

**Comment by tianocore-issues** [comment]
## Comment 19535
Date: 2022-09-08 12:01:59 +0000
From: @ardbiesheuvel

(In reply to Dionna Glaze from comment #20)
&gt; &gt; So is self-report kernel binary bit still an option?
&gt; 
&gt; It&apos;s unlikely, but maybe we can convince Linux upstream that it&apos;s really
&gt; necessary. There&apos;s a mechanism for it in the boot protocol called
&gt; kernel_info.
&gt; 
&gt; &gt; Shall we also consider the minimal amount of the memory that kernel needs. So that FW can accept that amount of memory then let OS handle the left unaccept-mem.
&gt; 
&gt; We could add a 64-bit entry in kernel_info called required_accepted_memory
&gt; and only have it be non-zero if the kernel supports unaccepted memory.
&gt; 
&gt; Ard, what do you think about this? I think since it involves necessary
&gt; information for booting (accepting the right amount of memory), this should
&gt; fit in the boot protocol.

I&apos;m skeptical but perhaps you could elaborate on the execution flow in this case? Who checks this bit and when/where? And how about other OSes or architectures? (note that &quot;we don&apos;t care&quot; is a valid response but it has to be explicit)

**Comment by tianocore-issues** [comment]
## Comment 19536
Date: 2022-09-08 12:20:11 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;


&gt; I&apos;m skeptical but perhaps you could elaborate on the execution flow in this case? Who checks this bit and when/where? And how about other OSes or architectures? (note that &quot;we don&apos;t care&quot; is a valid response but it has to be explicit)

We put the onus on bootloading modules at this point.

The idea is that either the Linux EFI module or the GRUB2 Linux bootloader will grab this data from the boot_params struct to create an event handler for ExitBootServices. If missing the required_accepted_memory entry, then accept all memory. If the entry exists, then accept enough memory to boot the OS to the point that the OS can interpret unaccepted_memory memory types from GetMemoryMap and accept as needed.

For direct boot from VMM (like CrosVM) I&apos;m little less familiar, but the idea would be that the VMM will use the boot_params information to determine how much memory should be part of the initially measured image at launch, as well as how to present the e820 map. This way even without EFI or bootloader to do the work, the kernel still has enough accepted memory to decompress itself to get to understanding the new memory type.

That&apos;s unfortunately many places to change since it counts as changing the boot protocol, but that&apos;s what I&apos;ve got for now.

**Comment by tianocore-issues** [comment]
## Comment 19548
Date: 2022-09-09 18:10:03 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt;Another question is that it seems the discussion is focused on whether the kernel support unaccept-mem or not. Shall we also consider the minimal amount of the memory that kernel needs. So that FW can accept that amount of memory then let OS handle the left unaccept-mem.

On the topic of reporting the required accepted memory amount, Min, do you know how one might calculate that for something like the linux kernel? What has TDX with lazy accept needed? If we add a kernel_info struct for unaccepted memory, then it can include things like absolute memory ranges that need to be accepted, and start and end address-relative address ranges that need to be accepted before StartImage. I&apos;ll need your input for the patch to Linux.

**Comment by tianocore-issues** [comment]
## Comment 19689
Date: 2022-10-03 12:52:45 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

# Document: UEFI Specification 2.10

# License

SPDX-License-Identifier: CC-BY-4.0

# Submitter: [TianoCore Community](https://www.tianocore.org)

# Summary of the change

Adds a new protocol to disable the UEFI&apos;s behavior of accepting all memory eagerly. 

# Benefits of the change

The new protocol allows a safe transition for operating systems to go from not supporting the EFI_UNACCEPTED_MEMORY type to then support it and have the UEFI do &quot;the right thing&quot;.

# Impact of the change

Confidential compute technologies that require memory acceptance will be able to boot much faster. This is due to two factors:

1. memory acceptance itself takes time
2. memory acceptance requires that the backing pages are pinned, which takes even more time.

Neither acceptance nor pinning for the entire virtual RAM space are required for boot.

Detailed description of the change [normative updates]

Changes to section 8.5.4: Exchanging information between the OS and Firmware

&quot;The firmware and an Operating System may exchange information through the OsIndicationsSupported and the
OSIndications variables as follows:&quot;

to become

&quot;The firmware and an Operating System may exchange information through protocols defined in [Protocols - OS to Firmware Communication](#chapter21) and the OsIndicationsSupported and the
OSIndications variables. The OsIndicationsSupported and OSIndications variable usage is as follows:&quot;

Addition of chapter between Chapter 20 Protocols - ACPI Protocols and Chapter 21 String Services, named Protocols - OS to Firmware communication

Introduction text:

&quot;There are situations such as a multi-boot system where a single system-wide bit in OsIndications is not suitable, and persistence is not necessary. In those cases, we specify protocols to communicate support for UEFI features that cannot be known ahead of time.&quot;

Section 21.1 - EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL

# Summary

# GUID:

```
#define EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL_GUID \
  {0xc5a010fe,0x38a7,0x4531, \
    {0x8a,0x4a,0x05,0x00,0xd2,0xfd,0x16,0x49}}


# Protocol Interface Structure:

```
typedef struct _EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL {
  EFI_DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY  Disable;
} EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL;
```

# Parameters:

# **Disable** Disables behavior of accepting all memory before ExitBootServices returns. See the [EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL.Disable()](#section21.1.1) function description.

# Description: This interface does not provide any way to re-enable the AcceptAllUnacceptedMemory behavior.

Section 21.1.1 EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL.Disable()

# Summary

Disables behavior of accepting all memory before ExitBootServices returns.

# Prototype:

```
typedef
EFI_STATUS
(EFIAPI *EFI_DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY)(
  IN  EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL  *This
  );
```

# Parameters:

**This** A pointer to EFI_ACCEPT_ALL_MEMORY_PROTOCOL instance. Type EFI_ACCEPT_ALL_MEMORY_PROTOCOL is defined in [EFI_ACCEPT_ALL_MEMORY_PROTOCOL](#section21.1)

# Description:

Meant to be called by OS loader if the OS supports EfiUnacceptedMemoryType and means to assume the role of accepting memory before its use. The UEFI is permitted to create unaccepted memory for the memory map whether or not this function is called. The intent is for the UEFI to default to the safe behavior of accepting all memory for the OS unless requests otherwise, but that is not required behavior.

==========================


Following Ard&apos;s advice, I tried the protocol implementation for disabling the default safe behavior of accepting all memory before ExitBootServices returns.
Do let me know if the means of accomplishing this disabling behavior must also be specified. For now you can see that I&apos;ve inserted a new callback point in DxeMain&apos;s CoreExitBootServices to be called after the timer is disabled and before TerminateMemoryMap is called.

I have a candidate implementation at https://github.com/deeglaze/edk2/pull/new/enable_umv5

**Comment by tianocore-issues** [comment]
## Comment 19800
Date: 2022-10-21 06:49:19 +0000
From: @jyao1

Thanks for the update. Some questions:

1) What is the difference between ACCEPT_ALL_UNACCEPTED_MEMORY and ACCEPT_ALL_MEMORY?

2) Does DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY == ENABLE_UNACCEPT_MEMORY ?
If yes, can we this simpler name?
If no, what is the difference?

3) Is this one permanent setting, then BIOS need save the configuration for future boot? Or one-time setting that only impact current boot?

**Comment by tianocore-issues** [comment]
## Comment 19801
Date: 2022-10-21 11:40:13 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; 1) What is the difference between ACCEPT_ALL_UNACCEPTED_MEMORY and ACCEPT_ALL_MEMORY?

ACCEPT_ALL_MEMORY is a typo. It should be ACCEPT_ALL_UNACCEPTED_MEMORY.


&gt;2) Does DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY == ENABLE_UNACCEPT_MEMORY ?
&gt;If yes, can we this simpler name?
&gt;If no, what is the difference?

This name is the result of an earlier comment from Ard on the EDK2 mailing list
https://edk2.groups.io/g/devel/message/94477?p=%2C%2C%2C20%2C0%2C0%2C0%3A%3Acreated%2C0%2CPcdEnableUnacceptedMemory+confusing%2C20%2C2%2C0%2C93975247

The answer is depends on the semantics the UEFI standards committee want to codify about unaccepted memory. If we say that the correct behavior of the UEFI is to always accept all unaccepted memory before ExitBootServices returns (exactly when is unspecified), then I&apos;m okay saying that we&apos;re &quot;enabling&quot; unaccepted memory.

If we&apos;re not saying that unaccepted memory should always be accepted without this protocol being called, then the specific name seems appropriate. If the UEFI has the behavior of accepting all unaccepted memory, then disable that behavior.

&gt; 3) Is this one permanent setting, then BIOS need save the configuration for future boot? Or one-time setting that only impact current boot?

This is a one-time setting for the current boot.

**Comment by tianocore-issues** [comment]
## Comment 19802
Date: 2022-10-21 12:20:24 +0000
From: @jyao1

&gt; ACCEPT_ALL_MEMORY is a typo. It should be ACCEPT_ALL_UNACCEPTED_MEMORY.

I think ACCEPT_ALL_MEMORY is enough.
Not sure why we need say ACCEPT_ALL_UNACCEPTED_MEMORY?
Is there any other memory we want to accept, such as ACCEPT_ALL_TYPEXXX_MEMORY?

&gt; Ard on the EDK2 mailing list

The comment is about PCD - PcdAcceptAllUnacceptedMemory v.s. PcdEnableUnacceptedMemory.
I agree that saying explicitly &quot;AcceptAll&quot; is better than &quot;enable&quot;

But saying: AcceptAllUnacceptedMemory seems redundant to me. Can we simply it to AcceptAllMemory?

Now back to the protocol/function name - EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL.EFI_DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY

I know it means what we want to do. But I just feel it is too hard to read it out, especially when the first word is DISABLE.

Can we make it simpler?
Protocol name candidate: EFI_MEMORY_ACCEPTING_PROTOCOL
Function name candidate: EFI_NOT_ACCEPT_ALL_MEMORY, or EFI_ALLOW_UNACCEPTED_MEMORY


&gt; This is a one-time setting for the current boot.

I feel we had better document the requirement in Description section clearly:
A) When the OS loader shall call this API, if supported.
B) When the BIOS shall accept all memory, if this is not called from OS loader.

Otherwise, the BIOS might accept all memory too early or the OS loader may call this API too late.

**Comment by tianocore-issues** [comment]
## Comment 19803
Date: 2022-10-21 12:49:07 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt;I think ACCEPT_ALL_MEMORY is enough.
&gt;Not sure why we need say ACCEPT_ALL_UNACCEPTED_MEMORY?

I say all unaccepted memory since we don&apos;t accept already accepted memory, since that would be a security vulnerability. If that&apos;s obvious enough from context, then I&apos;m fine dropping the _UNACCEPTED.

&gt;Is there any other memory we want to accept, such as ACCEPT_ALL_TYPEXXX_MEMORY?

There might be in the future? I don&apos;t know enough about the confidential PCI passthrough features to make bold claims that there won&apos;t be other kinds of memory. There could be DMA areas that need to be accepted and not become EfiConventionalMemory, but rather EfiMemoryMappedIO. I&apos;m not sure if the unaccepted memory type would have an attribute for that distinction, or if there would be an unaccepted_mmio type.

&gt;Now back to the protocol/function name - EFI_ACCEPT_ALL_UNACCEPTED_MEMORY_PROTOCOL.EFI_DISABLE_ACCEPT_ALL_UNACCEPTED_MEMORY

&gt;I know it means what we want to do. But I just feel it is too hard to read it out, especially when the first word is DISABLE.

&gt;Can we make it simpler?
&gt;Protocol name candidate: EFI_MEMORY_ACCEPTING_PROTOCOL
&gt;Function name candidate: EFI_NOT_ACCEPT_ALL_MEMORY, or EFI_ALLOW_UNACCEPTED_MEMORY

EFI_MEMORY_ACCEPTANCE_PROTOCOL seems more natural to me. I&apos;m down for EFI_ALLOW_UNACCEPTED_MEMORY.

&gt;I feel we had better document the requirement in Description section clearly:
&gt;A) When the OS loader shall call this API, if supported.
&gt;B) When the BIOS shall accept all memory, if this is not called from OS loader.
&gt;
&gt; Otherwise, the BIOS might accept all memory too early or the OS loader may call this API too late.

The BIOS might accept all memory because the firmware vendor decided that the runtime latency of accepting dynamically is too costly. I don&apos;t want to enforce that the BIOS /doesn&apos;t/ accept all memory, just that it&apos;s allowed to not accept all memory. We would count that as the OS supporting unaccepting memory and the firmware not supporting it.

In the other case, I agree that it is rather obtuse for the UEFI to ignore this protocol and produce unaccepted memory even when not called.

I had made a comment before that the OS loader has to call the protocol before ExitBootServices, but Ard pointed out that all protocols have that restriction, so saying it was redundant. It could be the first thing an OS loader does and that would still be safe, since the firmware has to ensure there&apos;s at least enough memory accepted for the image it loads.

diff:
 # Description
-Meant to be called by OS loader if the OS supports EfiUnacceptedMemoryType and means to assume the role of accepting memory before its use.
+The OS loader shall call this protocol if it supports the EfiUnacceptedMemoryType.
-The UEFI is permitted to create unaccepted memory for the memory map whether or not this function is called.
-The intent is for the UEFI to default to the safe behavior of accepting all memory for the OS unless requests otherwise, but that is not required behavior.
+If this protocol is not been called, then before ExitBootServices returns, the UEFI shall accept all EfiUnacceptedMemoryType regions, and subsequent calls to GetMemoryMap will report those regions as EfiConventionalMemory.

**Comment by tianocore-issues** [comment]
## Comment 19805
Date: 2022-10-21 22:41:57 +0000
From: @jyao1

&gt; DMA areas

Any DMA memory must be accepted, then converted from private to shared.

I don&apos;t think this protocol need care about DMA. It is out of scope.

&gt; EFI_MEMORY_ACCEPTANCE_PROTOCOL seems more natural to me. I&apos;m down for EFI_ALLOW_UNACCEPTED_MEMORY.

Thanks. That seems good to me.

&gt; The BIOS might accept all memory because the firmware vendor decided that the runtime latency of accepting dynamically is too costly. I don&apos;t want to enforce that the BIOS /doesn&apos;t/ accept all memory, just that it&apos;s allowed to not accept all memory. We would count that as the OS supporting unaccepting memory and the firmware not supporting it.

If that is the case, the BIOS shall NOT expose this protocol.

The implication is:
1) If BIOS expose this protocol, the OS makes decision and tell BIOS.
2) If the BIOS does not expose this protocol, the BIOS makes decision to accept all or not.

&gt; I had made a comment before that the OS loader has to call the protocol before ExitBootServices, but Ard pointed out that all protocols have that restriction, so saying it was redundant.

I dont think it is redundant. This sentence &quot;call the protocol before ExitBootServices&quot; implies that the BIOS must wait till ExitBootServices.

I provide another option: If supported, OS loader must locate this protocol and call this API immediately. Then the BIOS implementation may use a hook to see what is the first function call. If no, then BIOS can accept all much earlier.



Last but not least, we use a word document to track the CodeFirst content and use different version document to track the progress. We do not put all content to the comment section. If you can follow the same model, that would be great.

**Comment by tianocore-issues** [comment]
## Comment 19810
Date: 2022-10-24 13:38:29 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Created attachment 1040
Version 2 of the proposed spec change

Attachment: [BZ-1040-EFI_ACCEPT_ALL_UNACCEPTED_MEMORYv2.pdf](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1040-EFI_ACCEPT_ALL_UNACCEPTED_MEMORYv2.pdf)

**Comment by tianocore-issues** [comment]
## Comment 19811
Date: 2022-10-24 13:39:17 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Created attachment 1041
diff Version1 Version2

Attachment: [BZ-1041-EFI_ACCEPT_ALL_UNACCEPTED_MEMORYv2diff.pdf](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1041-EFI_ACCEPT_ALL_UNACCEPTED_MEMORYv2diff.pdf)

**Comment by tianocore-issues** [comment]
## Comment 19812
Date: 2022-10-24 13:40:19 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

I don&apos;t have word, so I hope pdf is acceptable.

**Comment by tianocore-issues** [comment]
## Comment 19813
Date: 2022-10-24 21:30:33 +0000
From: @jyao1

Thank you!

I have some other thought that I would like to discuss.

1) About mechanism.

I understand why that existing &quot;OsIndications&quot; cannot be used since it requires non-volatile storage (NV).

If we want to generalize the requirement, we just need define a way for &quot;one-time configuration from OS&quot;. One possible way is to define an &quot;OsOneTimeIndications&quot; volatile variable.

I dont mean we have to use variable. This is for discussion on what is best way to pass 1 bit configuration info with an extensible way.

2) About content location

If we finally decide to use EFI_MEMORY_ACCEPTANCE_PROTOCOL, I am thinking if it can be moved to &quot;Chapter 38 - Confidential Computing&quot;. This is a new chapter in UEFI 2.10, to describe the confidential computing related protocols.

The intention is to separate the confidential computing virtual BIOS specific thing from physical BIOS to avoid any misunderstanding.

People may argue that &quot;EfiUnacceptedMemoryType&quot; definition is not here. Because it is just a new memory type, it is naturally put with other memory type definition. We may consider to add more description there.

But if we decide to use &quot;OsOneTimeIndications&quot; variable, we had better keep the content in &quot;8.5.4. Exchanging information between the OS and Firmware&quot;.


3) About unaccepted memory size

In EDKII mailing list, we are discussing how the virtual BIOS knows how many memory it should accept before boot to OS which supports Unaccepted Memory.

Current decision is to unconditionally accept 4GiB, as known limitation/work-around.

I think this should be resolved in this protocol as well, as a complete end-to-end solution.

For example, when the OS/OsLoader says: &quot;I can allow the unaccepted memory&quot;, it should also say: &quot;I need xxxMB available memory at least, please ensure those size memory is accepted.&quot;

The possible solutions are:
A) Add one more parameter in AllowUnacceptedMemory(), such as &quot;UINT64  MinimalAcceptedMemorySize&quot;
B) Add several bits in &quot;OsOneTimeIndications&quot; variable, with 1MB granularity.

**Comment by tianocore-issues** [comment]
## Comment 19818
Date: 2022-10-25 12:21:21 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; I dont mean we have to use variable. This is for discussion on what is best way to pass 1 bit configuration info with an extensible way.

I think that OsIndications as a 64-bit variable with specified bit positions makes sense only because those bits need to be persisted. For something one-time each boot, I think that a UUID is better to identify a communication channel. It doesn&apos;t need to be a single bit either. The size could be determined by the UUID by convention, or made manifest in a configuration protocol that states the required data size for the UUID, and has a one-time message that takes that many bytes for the UUID.

We&apos;d then need a handle registration mechanism in the UEFI to allow registering handlers for data passed in from the OS.

This all seems like a meta-protocol, which I don&apos;t particularly like. I don&apos;t know why disparate communication channels need to be coalesced into one that has a constraint of one-time message sending. That constraint is not likely to have the versatility that some future problem&apos;s solution may need.

&gt; 2) Content location [...]

I&apos;m fine with wherever. I&apos;ll update the doc once we&apos;ve settled on a configuration protocol if it&apos;s not going to be the memory acceptance protocol.

&gt; In EDKII mailing list, we are discussing how the virtual BIOS knows how many memory it should accept before boot to OS which supports Unaccepted Memory.
&gt;Current decision is to unconditionally accept 4GiB, as known limitation/work-around.

Accept at most up to the 4GiB address (more likely 3GiB due to the MMIO hole) I think is more precise, but I understand. I don&apos;t know specifically how an OS would determine this number to communicate to the UEFI, so I couldn&apos;t really say whether accepting &lt; 4GiB is a good goal to try to achieve. Given the patches I&apos;ve seen that handle memory demand, it seems like a lot more complexity.

&gt;I think this should be resolved in this protocol as well, as a complete end-to-end solution.

Let me know when you have a proposal.

&gt; For example, when the OS/OsLoader says: &quot;I can allow the unaccepted memory&quot;, it should also say: &quot;I need xxxMB available memory at least, please ensure those size memory is accepted.&quot;

I don&apos;t know why this would need to be specified. The UEFI loads an image into memory that it has accepted and starts execution in it. The image must either call ExitBootServices and have the UEFI accept the memory, or accept memory itself. I&apos;m not familiar enough with how kernels manage their stack space in early boot to know if that&apos;s sufficient.

With something like a bootloader in the way, it sounds like the bootloader would need a binary format change from the OS image to know how much it should ask for on the OS&apos;s behalf, since the bootloader is also going to allocate memory for the next image. It can be done.. maybe, for Linux with the kernel_info boot parameter, but it seems awfully complicated.

&gt; The possible solutions are:
&gt;A) Add one more parameter in AllowUnacceptedMemory(), such as &quot;UINT64  MinimalAcceptedMemorySize&quot;

I&apos;d welcome a PoC to try, but I don&apos;t know how you imagine this working robustly.

&gt; B) Add several bits in &quot;OsOneTimeIndications&quot; variable, with 1MB granularity.

I would avoid a UEFI variable since they are R or R/W with no notion of one-time, and the bit positions would be limited and hard to add given the specification cadence. See reasoning above.

**Comment by tianocore-issues** [comment]
## Comment 19820
Date: 2022-10-25 13:23:36 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

If we&apos;re trying to make this protocol do more than I&apos;ve specified here, and there is no Code First use of it, I&apos;d much prefer that we leave that to a later specification.

If you want the flexibility of adding it as an entry point in this protocol, I could be down for having an entry point return the number version of the protocol so that it can be elaborated upon later.

Version 0 would have

EFI_STATUS EFIAPI AllowUnacceptedMemory(VOID) and
UINT32 EFIAPI Version(VOID);

Such that Version() returns 0 and the expected acceptance behavior is that up to address 4GiB is accepted before the first image is started.

I could imagine a version 1 that has another entry point like EFI_STATUS EFIAPI EnsureAccepted(EFI_PHYSICAL_ADDR StartAddress, UINTN NumPages) that a boot loader EFI driver could use to change the 4GiB acceptance behavior so that we have an OnReadyToBoot event that accepts up to 4GiB if EnsureAccepted has not been called.

But all that is to say, I&apos;d like for us to be able have this protocol while we&apos;re using the 4GiB solution and have an upgrade path for any new kind of feature.

**Comment by tianocore-issues** [comment]
## Comment 19824
Date: 2022-10-25 21:29:49 +0000
From: @jyao1

Let discuss MinimalAcceptedMemorySize at first.

4GiB is just one example, we think it may work in most cases. But there is no guarantee.

What if 4GiB is not enough?
What if we think 4GiB is too big and we want to save more boot time with 512MiB?

Do we want to say something like: If unaccepted memory is allowed, the BIOS shall accept 4GiB at least. This protocol consumer shall evaluate and ensure 4GiB accepted memory is large enough for setting up the reset memory acceptance action in OS. Otherwise, the protocol consumer shall not allow unaccepted memory.

I think your &quot;EnsureAccepted()&quot; can resolve those concern.



BTW: I dont think there is mandatory to implement all features to align with every version of spec. The final submission just need one version protocol.

**Comment by tianocore-issues** [comment]
## Comment 19844
Date: 2022-10-26 11:35:45 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; BTW: I dont think there is mandatory to implement all features to align with every version of spec. The final submission just need one version protocol.

I don&apos;t think this is true practically. OVMF still advertises itself as v2.7. If an OS finds a protocol that has one meaning at 2.7 and another meaning at 2.11, then the OS has to wait until the UEFI changes this very slow moving number. For a protocol to have its own version number, you can have much more granular control over what you can expect just from that protocol. That is why if we are considering adding to the protocol in a later specification, we should build versioning into its ABI on version 0.

**Comment by tianocore-issues** [comment]
## Comment 19890
Date: 2022-11-01 10:45:05 +0000
From: @jyao1

&gt; I don&apos;t think this is true practically. OVMF still advertises itself as v2.7. If an OS finds a protocol that has one meaning at 2.7 and another meaning at 2.11, then the OS has to wait until the UEFI changes this very slow moving number. For a protocol to have its own version number, you can have much more granular control over what you can expect just from that protocol. That is why if we are considering adding to the protocol in a later specification, we should build versioning into its ABI on version 0.

Sorry, I don&apos;t understand what you mean.

For version, I am talking about CodeFirst ECR proposal. There is no need to provide 2 different version.

To me, missing size is a gap for the CodeFirst ECR proposal.

**Comment by tianocore-issues** [comment]
## Comment 19911
Date: 2022-11-02 11:55:59 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt;For version, I am talking about CodeFirst ECR proposal. There is no need to provide 2 different version.
&gt;To me, missing size is a gap for the CodeFirst ECR proposal.

While I could certainly propose the EnsureAccepted method here, I don&apos;t have a plausible use for it in a Code First solution.

The problem in my understanding is that, let&apos;s say instead of accepting an arbitrary amount of 4GiB of memory, we instead accept only the memory that is used by the UEFI, including its allocations for loading images. How then can we ensure that

* when we start an image that the image will only touch accepted memory either until

   a. it requests the right amount of memory to make it to EBS without enabling unaccepted memory support, or

   b. until it enables unaccepted memory support and handles its own memory acceptance henceforth?

But the issue is that if the loaded image has no knowledge of the memory acceptance protocol, then neither a. nor b. are solutions. Instead we need a &quot;big enough&quot; initial acceptance that allows the image to make it to EBS for the UEFI to then accept all memory.

I don&apos;t see a compelling case to support a. and it doesn&apos;t seem like we can know we&apos;re in the b case or not until it&apos;s too late to accept more memory before starting the OS loader.

**Comment by tianocore-issues** [comment]
## Comment 19912
Date: 2022-11-02 12:06:32 +0000
From: @jyao1

&gt; Instead we need a &quot;big enough&quot; initial acceptance that allows the image to make it to EBS for the UEFI to then accept all memory.

So how can a BIOS know what value is &quot;big enough&quot; ?

**Comment by tianocore-issues** [comment]
## Comment 19916
Date: 2022-11-02 18:26:44 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; So how can a BIOS know what value is &quot;big enough&quot; ?

I think 4GiB is not entirely arbitrary. It&apos;s the entirety of 32-bit memory. The UEFI boot sequence sets up page tables only up to 4GiB, so we should be allowed to assume that the booted image will need to do something special before accessing higher memory. In that case they will also need to accept memory or have called EBS to have the UEFI accept all memory.

**Comment by tianocore-issues** [comment]
## Comment 19917
Date: 2022-11-02 19:59:42 +0000
From: @jyao1

I hope to use 128M or 256M if possible, especially in small environment.

Why we mandate 4G? 4G is too big in my opinion.

**Comment by tianocore-issues** [comment]
## Comment 19918
Date: 2022-11-02 20:26:19 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; I hope to use 128M or 256M if possible, especially in small environment.

If the environment is truly meant to be small, then the VM will not even have 4GiB attached. The acceptance is _up to_ 4GiB, provided that memory is present. If the environment is not meant to be small, then it can handle the milliseconds it takes to accept the 4GiB before starting the rest of the image to get to the EnableUnacceptedMemory method.

If the environment is not meant to be small and also isn&apos;t going to be enlightened to accept its own memory, then accepting 4GiB or before accepting the rest at EBS isn&apos;t going to make much difference over accepting 128MiB before EBS accepts &quot;the rest&quot; that includes the unaccepted regions with in the 4GiB - 128MiB lower memory range.

All of this isn&apos;t to say that the call for 4GiB memory acceptance is required, just required in the absence of a more specific protocol that could come later that is use case-driven. I don&apos;t have such a use case to drive a sound design.

This protocol would not preclude that possibility in the future. I just don&apos;t know what conditions would make that &quot;if possible&quot; a reality. Unless there&apos;s some binary inspection

&gt; Why we mandate 4G? 4G is too big in my opinion.

To support the OSes that don&apos;t understand unaccepted memory at all. That includes

1. The OS doesn&apos;t know to tell the UEFI that it needs at least X,Y,Z memory ranges accepted to get to EBS.

2. The OS boot protocol hasn&apos;t been updated to inform a boot loader of the required memory ranges to get to EBS.

3. The boot loader hasn&apos;t been updated to interpret an updated boot protocol to request that the UEFI accept the denoted memory ranges.

We mandate 4G because the UEFI sets up memory contents at both ends of the conventional memory ranges within 0..4GiB, and a launched image is free to operate normally with the assumption that memory mapped in the page table can be accessed without extra enlightenment. That assumption makes it possible for both enlightened and unenlightened images to execute all the way to ExitBootServices. At that time, the OS knows it needs to handle the rest of unaccepted memory, or it doesn&apos;t and doesn&apos;t have to because the memory has all been accepted.

Again I think that a later protocol or a later numbered version of this protocol should be what introduces the behavior you&apos;re asking for because it is much more complicated than simply saying that all memory mapped in the page table OnReadyToBoot should be accepted. I think that means all memory ranges within 0..4GiB.

**Comment by tianocore-issues** [comment]
## Comment 19919
Date: 2022-11-02 22:11:14 +0000
From: @jyao1

&gt; To support the OSes that don&apos;t understand unaccepted memory at all. 

&gt; That assumption makes it possible for both enlightened and unenlightened images to execute all the way to ExitBootServices. At that time, the OS knows it needs to handle the rest of unaccepted memory, or it doesn&apos;t and doesn&apos;t have to because the memory has all been accepted.


I don&apos;t understand this part.
If you are considering to support the OSes that don&apos;t understand unaccepted memory at all, then why we need introduce this new protocol?

The vBIOS can always accept 4GB and boot to OS. Done.

**Comment by tianocore-issues** [comment]
## Comment 19930
Date: 2022-11-03 12:40:07 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt;If you are considering to support the OSes that don&apos;t understand unaccepted memory at all, then why we need introduce this new protocol?
&gt;The vBIOS can always accept 4GB and boot to OS. Done.

Your Comment 27 called for specifying the memory acceptance behavior for when this protocol is not used.

We want UEFI to accept 4GB and boot to OS. What other option is there? UEFI doesn&apos;t accept 4GB for.. an unspecified reason, and then boots to the OS under the assumption that the OS will understand unaccepted memory?

How do we learn this reason early enough to /not/ accept the 4GiB of memory? This seems like the same problem we were trying to solve earlier in this bug just for enabling unaccepted memory. Ard showed me that we can delay acceptance until EBS to wait for an &quot;Enable&quot; call to disable the acceptance behavior.

But now say GRUB2&apos;s BOOTX64.EFI runs after being loaded. If it doesn&apos;t call EnsureAccepted for memory it needs, then the UEFI will have needed to accept the memory before loading BOOTX64.EFI. Oops. This is probably unfair, since GRUB2 will use AllocatePages, which Min Xu&apos;s lazy accept code would lazily accept before releasing it to GRUB2.

So GRUB2 has enough accepted memory to be able to parse the Linux boot parameters and learn that Linux needs memory ranges X0-Y0,X1-Y1 to be accepted. So GRUB2 calls EnsureAccepted for Linux&apos;s ranges and then continues boot. That might work to accept fewer than 4GB.

But what happens if GRUB2 doesn&apos;t see the boot parameters for Linux? It could force UEFI to accept up to 4GB to be safe.

What happens if GRUB2 hasn&apos;t been updated to know it needs to call EnsureAccepted though? There&apos;s no specific crossover call at which the UEFI would know, &quot;oh the bootloader doesn&apos;t understand restricted memory acceptance, so let&apos;s accept up to 4GB before continuing&quot;.

I&apos;m not seeing a safe enablement semantics for accepting fewer than 4GB.

**Comment by tianocore-issues** [comment]
## Comment 19935
Date: 2022-11-03 23:00:14 +0000
From: @jyao1

Hi Dionna
I am having a discussion with different teams and collecting data on how many memory acceptance is the best way to achieve boot performance.

I would be back to discuss more detail with solid data.

**Comment by tianocore-issues** [comment]
## Comment 19980
Date: 2022-11-09 14:42:23 +0000
From: @mdroth

Several comments:

You make reference to a EFI_MEMORY_ACCEPTANCE_PROTOCOL.Disable(), but there isn&apos;t one in the protocol definition.

The memory types are an enum, so theoretically if you&apos;re going to do an accept of a memory region that&apos;s decorated as a EfiUnacceptedMemoryType, what then does it become? How does this memory look like in the GUEST view of the memory map? EfiConventionalMemory? Whatever the answer, that needs to be documented as well.


Current language:

The UEFI shall accept all memory any time before ExitBootServices returns if it does not provide this protocol. The UEFI shall not accept all memory until ExitBootServices is called if the protocol is provided.

This protocol and acceptance behavior ensures that the UEFI safely supports OSes that don’t interpret the new EfiUnacceptedMemory type and those that do.


--- I don&apos;t understand the nuance you&apos;re grasping at here, because things triggered by ExitBootServices() aren&apos;t support to change the memory map, so ignore the whole ExitBootServices thing and describe what you want/problem statement - we can try to convert that to compliant language.

**Comment by tianocore-issues** [comment]
## Comment 19981
Date: 2022-11-09 15:40:43 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Created attachment 1051
Version 3 of the proposed spec change

Attachment: [BZ-1051-EFI_MEMORY_ACCEPTANCE_PROTOCOLv3.pdf](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1051-EFI_MEMORY_ACCEPTANCE_PROTOCOLv3.pdf)

**Comment by tianocore-issues** [comment]
## Comment 19982
Date: 2022-11-09 15:41:19 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Created attachment 1052
diff Version2 Version3

Attachment: [BZ-1052-EFI_MEMORY_ACCEPTANCE_PROTOCOLv3diff.pdf](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1052-EFI_MEMORY_ACCEPTANCE_PROTOCOLv3diff.pdf)

**Comment by tianocore-issues** [comment]
## Comment 19983
Date: 2022-11-09 15:47:42 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

Hi Michael, thanks for reviewing.

&gt; things triggered by ExitBootServices() aren&apos;t support to change the memory map, so ignore the whole ExitBootServices thing and describe what you want/problem statement - we can try to convert that to compliant language.

I&apos;ve updated the language in v3 to be more clear about what the protocol is trying to accomplish.

The ExitBootServices() call is a must-happen interaction with the UEFI that I use as a checkpoint that can accept all memory if the OS hasn&apos;t indicated support. Is there another such must-be-called function/protocol/service that could serve as this checkpoint? We need the OS to have an opportunity to run and advertise support, whereas we also need to be flexible enough such that any calls that a bootloader makes before loading the OS won&apos;t trigger the accept all behavior.

I thought we could make ExitBootServices return invalid_argument when we change the memory map to not have unaccepted memory anymore, and by spec, the caller needs to GetMemoryMap and re-call ExitBootServices. Ard Biesheuvel recommended this approach.

**Comment by tianocore-issues** [comment]
## Comment 20029
Date: 2022-11-15 21:40:16 +0000
From: @lgao4

Need spec change.

**Comment by tianocore-issues** [comment]
## Comment 20049
Date: 2022-11-16 12:07:24 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

&gt; Need spec change.

Are you asking for me to change something in my draft, or are you just confirming this &quot;Specification Update&quot; bug indeed needs a spec change?

**Comment by tianocore-issues** [comment]
## Comment 20163
Date: 2022-12-01 09:39:24 +0000
From: @jyao1

Feedback from Intel:

1) TDX virtual firmware (TDVF) will only support the guest OS which supports Lazy Accept, because of performance.

2) Intel Linux kernel maintainer has no plan to use this protocol in Linux as well. Reference: https://lore.kernel.org/all/bb7479df-7871-9861-600d-
c2fed783b659@intel.com/

3) All in all, we believe that a kernel without lazy accept is just short team. In long term, the kernel will support lazy accept by default.
Using policy protocol just for a short term is not preferred. It is useless in long term.
As such, TDVF and TDX-enlightened Guest will not use this policy protocol.

**Comment by tianocore-issues** [comment]
## Comment 20164
Date: 2022-12-01 11:47:56 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

This is a disappointing result, because it permanently destabilizes early adopters.

In GCP for example, we do not pin firmware versions. When a VM migrates, it receives the firmware that&apos;s released with the VMM that it has migrated to. This means a cold boot in a firmware that assumes lazy accept support for SEV-SNP will silently make most memory unavailable to the guest, only for them to experience an unexpected OOM at an unknown time in the future.

Ard Biesheuvel, another Linux EFI maintainer, has stated that he&apos;d accept a protocol use in Linux to enable lazy accept when the kernel supports it.

Proposed protocol: https://lkml.org/lkml/2022/6/27/990
Suggestion that he would accept it even without a spec change, although that is not a good outcome: https://lkml.org/lkml/2022/6/28/1240
Affirmative &quot;no problem adopting&quot; the use of the protocol: https://patchew.org/EDK2/20220928153323.2583389-1-dionnaglaze@google.com/20220928153323.2583389-6-dionnaglaze@google.com/#CAMj1kXFrsh83DT7PvrMtwWPO4OjkXt3E=Wc4Tis1Hbd+G._5FjaCQ@mail.gmail.com

Both TDX and SEV-SNP have proposed Kconfigs that make unaccepted memory optional, not required. This means any later version of the linux kernel can be configured in a breaking way due to the inability for a firmware to support both configurations of kernels.

**Comment by tianocore-issues** [comment]
## Comment 20970
Date: 2023-01-31 19:29:39 +0000
From: @jyao1

It becomes https://github.com/tianocore/edk2/blob/master/OvmfPkg/Include/Protocol/SevMemoryAcceptance.h.

Can we close this one?

**Comment by tianocore-issues** [comment]
## Comment 20971
Date: 2023-01-31 19:50:11 +0000
From: Dionna Glaze &lt;<dionnaglaze>&gt;

(In reply to jiewen.yao from comment #55)
&gt; It becomes
&gt; https://github.com/tianocore/edk2/blob/master/OvmfPkg/Include/Protocol/
&gt; SevMemoryAcceptance.h.
&gt; 
&gt; Can we close this one?

Yes, good to close.
